extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,150 @@
+        imageadElements[url]);
+  }
+  return imageadBatchManagers[url];
+}
+
+export class AmpAdBatchManager {
+
+  /**
+   * @param {string} url The base URL of the ad server
+   * @param {Array} elements An array of imagead Element objects to be batched together */
+  constructor(url, elements) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = url;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullUrl_ = this.url_;
+
+    /** @private {number} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.retries_ = AD_BATCH_RETRIES;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Get a list of slots that are in use for this URL
+    const slots = elements.map(element => {
+      const slotId = element.getAttribute('data-slot');
+      // Use a slot id of 0 if the element didn't provide one.
+      return slotId ? slotId : 0;
+    });
+    // If any slots were defined, add them to the URL
+    if (slots.length) {
+      this.fullUrl_ += '?s=' + slots.join(',');
+    }
+  }
+
+  /**
+   * Call this to instruct the batch manager to do the layout for an element. Prior to doing the layout, the batch manager
+   * will fetch the data (if this is the batch master) or wait for it to be ready (if not the batch master).
+   * @param {Element} elem The element for which the layout is to be done
+   * @returns {Promise} The element's "jsondata_" variable will be filled its with the data from the ad server before this is
+   *     resolved. If resolved OK, the element is returned by the promise
+   */
+  doLayout(elem) {
+    return new Promise(resolve => {
+      if (elem.batchMaster()) {
+        // This is the master. We're going to do the layout for all ads in the batch
+        // Gather the parameters to be added to the URL. First, make a list of the slot ids for all ads on this page with this URL
+        xhrFor(elem.win).fetchJson(this.fullUrl_).then(data => {
+          this.responseData = data;
+          resolve(elem);
+        }, function(err) {
+          user().error(TAG, err);
+        });
+      } else {
+        // This is not the master. We need to wait for the master to have fetched the response for the ad server;
+        //  when this is done,it's OK to proceed
+        this.getResponseWhenReady().then(function() {
+          resolve(elem);
+        }, function(err) {
+          user().error(TAG, err);
+        });
+      }
+    });
+  }
+
+  /**
+   * Return a promise that will complete when the response from the image server has been received.
+   * If the response is not ready after the specified timeout, reject the promise.
+   * @returns {Promise}
+   */
+  getResponseWhenReady() {
+    const t = this;
+    return new Promise(function(resolve) {
+      if (t.responseData === null) {
```

Comment:
I've now rewritten the whole thing to do something relatively similar to this. It simplified the code so much that the separate batch manager is no longer required. So I'm not going to reply individually to most of the comments on amp-ad-batch-manager.js.